### PR TITLE
feat(models): add CenterPoint

### PR DESCRIPTION
**Pull request by amadeuszsz**
## Summary

Add CenterPoint 3D object detection (`detection3d` task) with a PointPillars frontend, reusing the shared detection stack (`SECONDBackbone`, `SECONDFPN`, `CenterHead`).

- Model: `CenterPointDetectionModel` wires `PillarFeatureNet` + `PointPillarsScatter` + `SECONDBackbone` + `SECONDFPN` + the existing `CenterHead` (velocity regression enabled).
- Input: 5 point features `(x, y, z, intensity, time_lag)` — `LoadPointsFromMultiSweeps` with `time_dim` fills the per-point time lag (T4: key frame + 1 sweep, NuScenes: 10 frames).
- `PillarFeatureNet` decorates pillars with cluster-center and 3D voxel-center offsets — an 11-feature layout identical to the `autoware_lidar_centerpoint` preprocess kernel (`POINT_DIM_XYZIT`, `encoder_in_feature_size: 11`).
- Configs: NuScenes 51.2 m and T4Dataset (J6Gen2) 120 m, matching the deployed AWML CenterPoint parameters (`feat_channels: [32, 32]`, 32 points/voxel, 96000 voxels, `out_size_factor: 4`/`out_size_factor: 2`).
- Deployment: two-module ONNX split consumed by `autoware_universe/perception/autoware_lidar_centerpoint` — `pts_voxel_encoder_centerpoint.onnx` (decorated pillar features → per-pillar descriptors) and `pts_backbone_neck_head_centerpoint.onnx` (BEV canvas → raw `heatmap` / `reg` / `height` / `dim` / `rot` / `vel`). Voxelization, pillar decoration, BEV scatter, and box decoding run in the runtime node.

Note: stacked on #67 - the multi-sweep `time_lag` loading and key-frame timestamp propagation used by the configs land there.

## Checklist

- [ ] You've checked our [contribution overview](https://tier4.github.io/autoware-ml/contributing/contribution-overview/).
- [ ] Your PR follows our [pull request guidelines](https://tier4.github.io/autoware-ml/contributing/pull-request-guidelines/).

### CODE REVIEW

File: autoware_ml/configs/tasks/detection3d/centerpoint/voxel020_second_secfpn_51m_nuscenes.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  train_transforms:
+    _target_: autoware_ml.transforms.base.TransformsCompose
+    pipeline:
+      - _target_: autoware_ml.transforms.boxes3d.loading.LoadAnnotations3D
+        name_mapping: ${dataset.detection3d.name_mapping}
+      - _target_: autoware_ml.transforms.point_cloud.sweeps.LoadPointsFromMultiSweeps
+        sweeps_num: 10
+        load_dim: 5
+        use_dim: [0, 1, 2, 3, 4]
+        time_dim: 4
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/configs/tasks/detection3d/centerpoint/voxel020_second_secfpn_51m_nuscenes.yaml:53]: Linked PR https://github.com/tier4/autoware-ml/pull/67 comes with required change

File: autoware_ml/configs/tasks/detection3d/centerpoint/voxel020_second_secfpn_51m_nuscenes.yaml [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+voxel_size: [0.2, 0.2, 8.0]
+
+model:
+  pts_middle_encoder:
+    output_shape: [512, 512]
+  pts_backbone:
+    layer_strides: [2, 2, 2]
+  bbox_head:
+    out_size_factor: 4
+    nms_min_radius: 1.0
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/configs/tasks/detection3d/centerpoint/voxel020_second_secfpn_51m_nuscenes.yaml:23]: Adjusted description

File: autoware_ml/models/detection3d/centerpoint.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            voxel_coords: Batched voxel coordinates.
+            batch_size: Optional explicit batch size.
+
+        Returns:
+            Detection head outputs.
+        """
+        if batch_size is None:
+            batch_size = infer_batch_size_from_voxel_coords(voxel_coords)
+        point_features = self.pts_voxel_encoder(voxels, num_points, voxel_coords)
+        bev_features = self.pts_middle_encoder(point_features, voxel_coords, batch_size=batch_size)
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/models/detection3d/centerpoint.py]: Out of scope.

File: autoware_ml/models/detection3d/centerpoint.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+                module=_CenterPointBackboneNeckHeadExportWrapper(
+                    self.pts_backbone,
+                    self.pts_neck,
+                    self.bbox_head,
+                ),
+                args=(spatial_features,),
+                input_param_names=["spatial_features"],
+                output_names=_CenterPointBackboneNeckHeadExportWrapper.output_names,
+            ),
+        }
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/models/detection3d/centerpoint.py:203]: https://github.com/tier4/autoware-ml/pull/71/commits/cce16222ddc3c97eed1093cded36d39b8951ebbd

File: autoware_ml/models/detection3d/centerpoint.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        self.backbone = backbone
+        self.neck = neck
+        self.bbox_head = bbox_head.prepare_for_export()
+
+    def forward(self, spatial_features: torch.Tensor) -> tuple[torch.Tensor, ...]:
+        """Run CenterPoint BEV feature extraction and prediction heads."""
+        bev_features = self.backbone(spatial_features)
+        bev_features = self.neck(bev_features)
+        outputs = self.bbox_head(bev_features)
+        return tuple(outputs[name] for name in self.output_names)
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/models/detection3d/centerpoint.py:70]: https://github.com/tier4/autoware-ml/pull/71/commits/cce16222ddc3c97eed1093cded36d39b8951ebbd

File: autoware_ml/models/detection3d/centerpoint.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    def __init__(
+        self,
+        pts_voxel_encoder: torch.nn.Module,
+        pts_middle_encoder: torch.nn.Module,
+        pts_backbone: torch.nn.Module,
+        pts_neck: torch.nn.Module,
+        bbox_head: torch.nn.Module,
+        optimizer: Callable[..., Optimizer] | None = None,
+        scheduler: Callable[[Optimizer], LRScheduler] | None = None,
+        metrics: list[MetricSuite] | None = None,
```
Comment (KSeangTan) [inline autoware_ml/models/detection3d/centerpoint.py]: Use `Sequence` instead of `list`

File: autoware_ml/models/detection3d/centerpoint.py [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+        super().__init__(optimizer=optimizer, scheduler=scheduler, metrics=metrics)
+        self.pts_voxel_encoder = pts_voxel_encoder
+        self.pts_middle_encoder = pts_middle_encoder
+        self.pts_backbone = pts_backbone
+        self.pts_neck = pts_neck
+        self.bbox_head = bbox_head
+
+    def build_eval_output(self, batch: Mapping[str, Any], outputs: Any) -> dict[str, Any]:
+        """Decode detections and pair them with ground truth for metrics."""
+        return detection_eval_output(self.bbox_head.predict(outputs), batch)
```
Comment (KSeangTan) [inline autoware_ml/models/detection3d/centerpoint.py:112]: We should likely make the function:
`detection_eval_output` in OOP to support all detection model class right?

File: autoware_ml/models/detection3d/centerpoint.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+    def _forward_with_batch_size(
```
Comment (KSeangTan) [inline autoware_ml/models/detection3d/centerpoint.py]: Explain more what does `forward_with_batch_size` mean?

File: autoware_ml/models/detection3d/centerpoint.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+
+        Args:
+            voxels: Voxel features.
+            num_points: Number of points in each voxel.
+            voxel_coords: Batched voxel coordinates.
+
+        Returns:
+            Detection head outputs.
+        """
+        return self._forward_with_batch_size(
```
Comment (KSeangTan) [inline autoware_ml/models/detection3d/centerpoint.py]: Any reason why dont we simply write the implementation here?

File: autoware_ml/models/detection3d/centerpoint.py [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+            voxels=voxels, num_points=num_points, voxel_coords=voxel_coords
+        )
+
+    def compute_metrics(
```
Comment (KSeangTan) [inline autoware_ml/models/detection3d/centerpoint.py:137]: Should be compute_loss instead of compute_metrics?

File: autoware_ml/models/detection3d/centerpoint.py [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        """Compute CenterPoint training losses."""
+        return self.bbox_head.loss(
+            outputs, batch_inputs_dict["gt_boxes"], batch_inputs_dict["gt_labels"]
+        )
+
+    def predict_outputs(
+        self, batch_inputs_dict: dict[str, Any], outputs: dict[str, torch.Tensor]
+    ) -> Any:
+        """Decode predictions for inference."""
+        del batch_inputs_dict
```
Comment (KSeangTan) [inline autoware_ml/models/detection3d/centerpoint.py:151]: IMO, we don't actually need to explicitly write
```
del batch_inputs_dict
```
We should redesign it and remove this from the function interface.

File: autoware_ml/models/detection3d/centerpoint.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+        return self.bbox_head.predict(outputs)
+
+    def get_log_batch_size(self, batch_inputs_dict: dict[str, Any]) -> int | None:
+        """Log the sample count instead of voxel count for lidar detection."""
+        return len(batch_inputs_dict["gt_boxes"])
+
+    def build_export_spec(self, batch_inputs_dict: Mapping[str, Any]) -> ExportSpec:
```
Comment (KSeangTan) [inline autoware_ml/models/detection3d/centerpoint.py:158]: TODO:
export-related or deployment-related should be done in another class to make the class cleaner.

**Review by KSeangTan — APPROVED** [review]
LGTM overall, i think some of them need to be redesigned to make the framework cleaner. Let's proceed with merging first.

File: autoware_ml/models/detection3d/centerpoint.py (hunk 6 above)
Comment (amadeuszsz) [reply autoware_ml/models/detection3d/centerpoint.py]: Thanks, https://github.com/tier4/autoware-ml/pull/71/commits/f195ee9a567ac3b2e10271a8833ab72a9682e9c9

File: autoware_ml/models/detection3d/centerpoint.py (hunk 8 above)
Comment (amadeuszsz) [reply autoware_ml/models/detection3d/centerpoint.py]: Thanks. Here we don't need it, the concept comes from different model. https://github.com/tier4/autoware-ml/pull/71/commits/2fe39c7636f474fd6e96ee613a0cf462cffb1b60

File: autoware_ml/models/detection3d/centerpoint.py (hunk 9 above)
Comment (amadeuszsz) [reply autoware_ml/models/detection3d/centerpoint.py]: Thanks. Here we don't need it, the concept comes from different model. https://github.com/tier4/autoware-ml/pull/71/commits/2fe39c7636f474fd6e96ee613a0cf462cffb1b60

File: autoware_ml/models/detection3d/centerpoint.py (hunk 10 above)
Comment (amadeuszsz) [reply autoware_ml/models/detection3d/centerpoint.py:137]: Name come from base class. It can be misleading. I need to update base class and potentially split the function to compute_loss and compute_metrics.

File: autoware_ml/models/detection3d/centerpoint.py (hunk 11 above)
Comment (amadeuszsz) [reply autoware_ml/models/detection3d/centerpoint.py:151]: This comes from base class signature as some model needs `batch_inputs_dict` here. I will look for a solution where we can drop `batch_inputs_dict` in `predict_outputs` method signature.

File: autoware_ml/models/detection3d/centerpoint.py (hunk 7 above)
Comment (amadeuszsz) [reply autoware_ml/models/detection3d/centerpoint.py:112]: Happy to address after StreamPETR integration is done (easier to build strong abstraction).
